src/utils/printing/getPrinterIppAttributes.ts

```diff
@@ -0,0 +1,261 @@
+        level,
+      } as IppMarkerInfo),
+  )
+
+  return { state, stateReasons, markerInfos }
+}
+
+/**
+ * Parse the output of ipptool. It looks like this:
+ *  "/dev/stdin":
+ *      Get-Printer-Attributes:
+ *          attributes-charset (charset) = utf-8
+ *          attributes-natural-language (naturalLanguage) = en
+ *          printer-uri (uri) = ipp://localhost:60000/ipp/print
+ *          requested-attributes (1setOf keyword) = printer-state,printer-state-reasons
+ *      /dev/stdin                                                           [PASS]
+ *          RECEIVED: 183 bytes in response
+ *          status-code = successful-ok (successful-ok)
+ *          attributes-charset (charset) = utf-8
+ *          attributes-natural-language (naturalLanguage) = en
+ *          printer-state (enum) = stopped
+ *          printer-state-reasons (1setOf keyword) = media-empty-error,media-needed-error,media-empty-error
+ */
+type IppAttributes = {
+  [attribute: string]: string | string[] | number | number[]
+}
+function parseIpptoolOutput(output: string): IppAttributes {
+  debug('parsing ipptool output: %s', output)
+  let lines = output.split('\n')
+  assert(lines.shift() === '"/dev/stdin":')
+  lines = lines.slice(
+    lines.findIndex(line => line.trim().startsWith('/dev/stdin')) + 1,
+  )
+  assert(
+    lines
+      .shift()
+      ?.trim()
+      .startsWith('RECEIVED:'),
+  )
+  const statusLine = lines.shift()?.trim()
+  if (statusLine !== 'status-code = successful-ok (successful-ok)') {
+    throw Error(`Unsuccessful ipptool response: ${statusLine}`)
+  }
+  assert(lines.shift()?.trim() === 'attributes-charset (charset) = utf-8')
+  assert(
+    lines.shift()?.trim() ===
+      'attributes-natural-language (naturalLanguage) = en',
+  )
+
+  const lineRegex = /^\s*(.+) \((.+)\) = (.*)$/
+  const attributes = Object.fromEntries(
+    lines
+      .filter(line => line !== '')
+      .map(line => {
+        const matches = lineRegex.exec(line)
+        if (!matches) {
+          throw Error(`Unable to parse ipptool output line: ${line}`)
+        }
+        const [, attribute, type, value] = matches
+        switch (type) {
+          case 'keyword':
+          case 'enum':
+          case 'nameWithoutLanguage':
+            return [attribute, value]
+          case 'integer':
+            return [attribute, parseInt(value)]
+          case '1setOf keyword':
+          case '1setOf enum':
+          case '1setOf nameWithoutLanguage':
+            return [attribute, value.split(',')]
+          case '1setOf integer':
+            return [attribute, value.split(',').map(parseInt)]
+          default:
+            throw Error(`Unrecognized ipptool attribute type: ${type}`)
+        }
+      }),
+  )
+  debug('parsed ipptool attributes: %O', attributes)
+  return attributes
+}
+
+function wrapWithArray<T>(value: T | T[]): T[] {
+  return Array.isArray(value) ? value : [value]
+}
+
+function zip(...arrays: unknown[][]): unknown[][] {
```

Comment:
You could copy  the version of this from vxsuite to get better types.

Reply:
funny enough, vxsuite only has types for up to five arrays, and here we need six. I considered doing an explicit type like that, but even with that type I'd still need type assertions to specify which attributes are numbers and which are strings, so i don't know that there's a huge benefit